src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,312 @@
+                    // for this project.  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
+            }
+            
+            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            {
+                return UpdateSymbolTreeInfoAsync(project, cancellationToken);
+            }
+
+            private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
                 {
-                    if (!_symbolCountByProjectMap.ContainsKey(projectId))
-                    {
-                        return SpecializedTasks.EmptyTask;
-                    }
+                    return;
+                }
+
+                Project lastSeenProject;
+                if (_idTolastSeenProject.TryGetValue(project.Id, out lastSeenProject) && lastSeenProject == project)
+                {
+                    // We already saw this project.  No need to do anything;
+                    return;
                 }
 
-                ReportCount();
-                return SpecializedTasks.EmptyTask;
+                // Check the semantic version of this project.  The semantic version will change
+                // if any of the source files changed, or if the project version itself changed.
+                // (The latter happens when something happens to the project like metadata 
+                // changing on disk).
+                var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+                ProjectInfo projectInfo;
+                if (!_projectToInfo.TryGetValue(project.Id, out projectInfo) || projectInfo.VersionStamp != version)
+                {
+                    await UpdateReferencesAync(project, cancellationToken).ConfigureAwait(false);
+
+                    var info = await SymbolTreeInfo.GetInfoForSourceAssemblyAsync(project, cancellationToken).ConfigureAwait(false);
+                    projectInfo = new ProjectInfo(version, info);
+                    _projectToInfo.AddOrUpdate(project.Id, projectInfo, (_1, _2) => projectInfo);
+                }
+
+                // Mark that we've completed processing this project.
+                _idTolastSeenProject[project.Id] = project;
             }
 
-            public override void RemoveProject(ProjectId projectId)
+            private async Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
             {
-                if (_symbolCountByProjectMap != null)
+                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
```

Comment:
I think you probably want to play with multiple version. for this, I think you want Project.Version.